test/expect-tests/memo/memoize_tests.ml

```diff
@@ -1260,6 +1260,120 @@ let%expect_test "Nested nodes with cutoff are recomputed unnecessarily (async)"
+      ~doc:"" Async
+      (fun offset ->
+        printf "Started evaluating summit\n";
+        let middle =
+          create ~with_cutoff:false "middle" (fun () ->
+              printf "Started evaluating middle\n";
+              incr counter;
+              let+ result =
+                match !counter with
+                | 1 ->
+                  printf "*** Depending on the cell ***\n";
+                  Memo.Cell.get_async cell
+                | _ ->
+                  printf "*** Not depending on the cell ***\n";
+                  Build.return 0
+              in
+              printf "Evaluated middle: %d\n" result;
+              result)
+        in
+        let+ middle = Memo.exec middle () in
+        let result = middle + offset in
+        printf "Evaluated summit: %d\n" result;
+        result)
+  in
+  evaluate_and_print summit 0;
+  [%expect
+    {|
+    Started evaluating summit
+    Started evaluating middle
+    *** Depending on the cell ***
+    Started evaluating base
+    Evaluated base: 8
+    Evaluated middle: 8
+    Evaluated summit: 8
+    f 0 = Ok 8
+    |}];
+  Memo.restart_current_run ();
+  evaluate_and_print summit 0;
+  (* No recomputation is needed since the [cell] is up to date. *)
+  [%expect {| f 0 = Ok 8 |}];
+  Memo.Cell.invalidate cell;
+  Memo.restart_current_run ();
+  evaluate_and_print summit 0;
+  (* Note that we no longer depend on the [cell]. The corresponding message is
+     printed twice due to the known performance issue with nested nodes. *)
+  [%expect
+    {|
+    Started evaluating base
+    Evaluated base: 8
+    Started evaluating middle
+    *** Not depending on the cell ***
+    Evaluated middle: 0
+    Started evaluating summit
+    Started evaluating middle
+    *** Not depending on the cell ***
+    Evaluated middle: 0
+    Evaluated summit: 0
+    f 0 = Ok 0 |}];
+  Memo.Cell.invalidate cell;
+  Memo.restart_current_run ();
+  evaluate_and_print summit 0;
+  (* The output below is unexpected: the cell is no longer a dependency of the
+     [middle] node yet it still somehow invalidated the result. The reason is
+     that the cell is a "phantom" dependency: we started checking whether the
+     dependencies of [middle] were up to date, found that they were not, but
+     still registered the dependency, and then recomputed [middle], keeping the
+     phantom dependency. *)
+  [%expect
+    {|
+    Started evaluating base
+    Evaluated base: 8
+    Started evaluating middle
+    *** Not depending on the cell ***
+    Evaluated middle: 0
+    Started evaluating summit
+    Started evaluating middle
+    *** Not depending on the cell ***
+    Evaluated middle: 0
+    Evaluated summit: 0
+    f 0 = Ok 0 |}];
+  Memo.Cell.invalidate cell;
+  Memo.restart_current_run ();
+  evaluate_and_print summit 0;
+  (* Note that phantom dependencies are persistent: we keep re-checking and
+     re-registering them. *)
+  [%expect
+    {|
+    Started evaluating base
```

Comment:
"Evaluated base" is the problem here, right?
I feel like we should strive to print less, otherwise it's hard to see the relevant things among the noise.
Maybe we could introduce a [debug : bool ref] somewhere (maybe even in the memo library itself) to make these printouts easy to enable only when debugging.

Reply:
Agreed. Some debugging infrastructure in the Memo would be very helpful. For now, I'll reduce noise in this particular test.